Separate dnschallenge executable

samgiles opened the pull request:
Same as #376 

Testin' Travis.

samgiles commented:
r? @fabricedesre

It seems the OpenSSL link failures on Mac OSX are intermittent - although it doesn't affect this PR :( - maybe because of cached binaries?

fabricedesre commented:
Any reason for the dnschallenge program to not be in its own crate (even in the same repo) ? It looks self contained enough.

samgiles commented:
That would be nice, especially if it was in the same repo - the reason it's not here is because the `tls` stuff isn't in a separate crate which it depends on.  It would be cool to move that out too, maybe to a `net` crate.  I could give this a go for this PR.

samgiles commented:
@fabricedesre how does this look?

@JohanLorenzo  Here I've made a components directory with separate crates, and updated the test scripts to run the tests in each components directory.  I've also added to build.rs so that the dns challenge binary is built when running `cargo build`.  How does this look?

fabricedesre commented on README.md:
```diff
@@ -91,6 +91,19 @@ default = []
+| `bash`         | No (required for LetsEncrypt client)           | System package manager                                                 |
+| `python`       | Yes (required to run `pagekite.py`)            | System package manager                                                 |
+| `pagekite.py`  | Yes (required to enable tunneling)             | https://pagekite.net/wiki/OpenSource/                                  |
+
```
nit: not sure if this is because of the diff, but we may have a alignment issues in this table. My eyes are bleeding.

fabricedesre commented on components/dns_challenge/main.rs:
```diff
@@ -0,0 +1,72 @@
+
+#![allow(dead_code)]
+
+// Needed to derive `Serialize` on ServiceProperties
```
nit: invalid comment. (we need to update the main src/main.rs too)

fabricedesre commented on components/dns_challenge/main.rs:
```diff
@@ -0,0 +1,72 @@
+// Needed for many #[derive(...)] macros
+#![allow(used_underscore_binding)]
+
+#![feature(associated_consts)]
```
can you double check what you need exactly? It looks like `#![allow(used_underscore_binding)]` and `#![feature(associated_consts)]` may not be needed

fabricedesre commented on components/dns_challenge/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+openssl-sys = "0.7.6"
+serde = "0.7.0"
+serde_json = "0.7.0"
+serde_macros = "0.7.2"
```
why do we need serde?

fabricedesre commented on components/tls/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+openssl-sys = "0.7.6"
+serde = "0.7.0"
+serde_json = "0.7.0"
+serde_macros = "0.7.2"
```
same here: do we need serde?

fabricedesre commented:
I haven't tested, but overall I like it! @JohanLorenzo that would mean that thinkerbell would move to components/thinkerbell too.

JohanLorenzo commented on tools/travis-linux.sh:
```diff
@@ -43,7 +43,16 @@ set_up_tests() {
     # Note: Cargo recompiles every dependency with dead code. That's why
     # this step is currently the last
+
+    echo "run_tests: Running crate tests..."
```
FYI: I [centralized that logic](https://github.com/fxbox/foxbox/pull/406/files#diff-f4983c80cbd9b1e076492899a26bd4d8R11) in #406. 

As rebasing #406 might be tricky (due to commits coming from another repo), would you mind waiting on that PR to land?

JohanLorenzo commented on build.rs:
```diff
@@ -44,7 +46,21 @@ fn copy_shared_static_files() {
 fn main() {
     update_local_git_hook();
     copy_shared_static_files();
+    cargo_build_in_directory("./components/dns_challenge").unwrap();
```
LGTM

JohanLorenzo commented on tools/travis-linux.sh:
```diff
@@ -43,7 +43,16 @@ set_up_tests() {
     # this step is currently the last
+
+    echo "run_tests: Running crate tests..."
+    for DIR in $(ls -d components/*); do
```
It's usually a [bad practice](http://mywiki.wooledge.org/BashGuide/Practices#Don.27t_Ever_Do_These) to use the output of `ls` mainly because of spaces in file names. `find` would be more reliable

samgiles commented:
@fabricedesre Turns out most of the stuff isn't needed anymore in the `dnschallenge` executable - only the `tls` crate. 
The formatting was a little off on one line in the README, a `|` wasn't extended to the edge of the table.  But, yeh, the diff looks horrible because of wrapping, it looks alright in an editor and renders the right thing.

Ok to merge?

fabricedesre commented:
Thanks, go ahead!
